core/CronArchive/QueueConsumer.php

```diff
@@ -201,6 +201,13 @@ public function getNextArchivesToProcess()
                 continue;
             }
 
+            if ($this->model->isSimilarArchiveInProgress($invalidatedArchive)) {
```

Comment:
@diosmosis I now wonder if things need to be more complicated... like do we need to know the "Archive start time" of that archive that is already in progress and only cancel invalidations that were added to the invalidations table before the start time? 

Eg 
* add invalidation 1 for Month Sep 2020 at 7AM
* add invalidation 2 for Month Sep 2020 at 8AM
* add invalidation 3 for Month Sep 2020 at 9AM
* start archiving Month Sep 2020 at 09:30 AM
* add invalidation 4 for Month Sep 2020 at 10AM
* add invalidation 5 for Month Sep 2020 at 11AM
* archive finishes... we should now remove the IDs 1,2,3 but maybe keep 4 and 5 or at least only 5?
* Meaning 1,2,3 would be removed, and 4 and 5 would be just skipped here?

Not sure if that's possible or if I'm missing something.

This way we make sure to start an archive again after that one finished and include all data.

Reply:
This is what would currently happen:

* add invalidation 1 for Month Sep 2020 at 7AM
* add invalidation 2 for Month Sep 2020 at 8AM
* add invalidation 3 for Month Sep 2020 at 9AM
* start archiving Month Sep 2020 at 09:30 AM
* core:archive picks up invalidation 2, tries to lock or sees that there is an existing one and ignores it
* core:archive picks up invalidation 3, tries to lock or sees that there is an existing one and ignores it
* something adds invalidation 4 for Month Sep 2020 at 10AM
* something add invalidation 5 for Month Sep 2020 at 11AM
* archive finishes... we should now remove the IDs 1,2,3 but maybe keep 4 and 5 or at least only 5?
* core:archive pulls the next archive, sees invalidation 2. it checks the archive's ts_archived, sees that it was finished moments ago and ignores it but does not delete it. (see QueueConsumer::usableArchiveExists() and where it is called).
* same thing happens for 2-5
* next hourly cron, invalidation 2 is picked up. and the process repeats.

w/ this PR:

* add invalidation 1 for Month Sep 2020 at 7AM
* add invalidation 2 for Month Sep 2020 at 8AM
* add invalidation 3 for Month Sep 2020 at 9AM
* start archiving Month Sep 2020 at 09:30 AM
* core:archive picks up invalidation 2, tries to lock or sees that there is an existing one and discards it
* core:archive picks up invalidation 3, tries to lock or sees that there is an existing one and discards it
* something adds invalidation 4 for Month Sep 2020 at 10AM
* something add invalidation 5 for Month Sep 2020 at 11AM
* archive finishes... we should now remove the IDs 1,2,3 but maybe keep 4 and 5 or at least only 5?
* core:archive pulls the next archive, sees invalidation 4. there is none in progress, so it passes that check. then it checks the archive's ts_archived, sees that it was finished moments ago and ignores it but does not delete it. (see QueueConsumer::usableArchiveExists() and where it is called).
* same thing happens to invalidation 5.
* next hourly cron, invalidation 4 is picked up. it is started.
* invalidation 5 is seen, since invalidation 4 is in progress, it is discarded.
* invalidation 4 is finished and there are no more invalidations left.

I don't think involving time is needed.

Reply:
Just to be sure I'm fully understanding it @diosmosis 

If there are multiple archivers launching at the very same time (which we support), or if there are more archivers starting between 9:30am (start of the archive) and 11:30am (finish of archive). Would one of the other archiving jobs maybe try to process invalidation 4 and 5 and remove them?

Reply:
@tsteur if you mean multiple core:archive processes, then no, because we still use the site ID queue (ie, SharedSiteIds), so we shouldn't be archiving the same site at the same time.

Reply:
I'm not sure if it's still the case but I assume it might be. What would happen is that when all sites have been processed, the queue would reset and all sites would be added again. For simplicity lets say there's only one site. The 1 site would be taken out of the queue thus the queue be then empty. The next time the cron runs it would add idSite=1 again. So they might run in parallel.

Reply:
Maybe we could use a Lock per site w/ 12-24 hour timeout or so that is checked before starting a site? Expired every hour at most. Or something like that. Or maybe just set an option saying site N is being archived, though safer to use a lock since it has an expiry time.

Seems like it would be easier to reason about the code if we make sure the same site is never archived concurrently.

Reply:
It should be definitely possible to archive the same site concurrently. AFAIK many users do that in order to process data faster and it is sometimes needed for big sites. I'd definitely prefer not to use any locks. 

Maybe instead of using archive_start time we could use the `idinvalidation`? Eg if there is a similar archive in progress, and the current ID is higher than the in progress ID then we simply skip it. If the ID is lower than we skip and discard?

Reply:
> It should be definitely possible to archive the same site concurrently. 

I thought the whole point of SharedSiteIds/FixedSiteIds was to avoid this?

> Maybe instead of using archive_start time we could use the idinvalidation? Eg if there is a similar archive in progress, and the current ID is higher than the in progress ID then we simply skip it. If the ID is lower than we skip and discard?

Sure, that sounds doable.

Reply:
> I thought the whole point of SharedSiteIds/FixedSiteIds was to avoid this?

I think it mostly makes sure we rotate through each site (especially should there we a lot of sites and few sites are taking a long time so other sites would otherwise not get a chance to archive data).

Reply:
> It should be definitely possible to archive the same site concurrently.

Btw was quickly reading and noticed this comment, I don't think that the case? (or it would be a new feature maybe). afaik the `core:archive` will skip a website ID when a website ID is currently being archived. It will only allow 1 thread per website even if the website has many segments.

Reply:
You can customise how many concurrent are possible AFAIK, and it should allow eg archiving different days in parallel etc

Reply:
Looking at the code it looks like it's this isn't possible? Can you verify this is how it works in SharedSiteIds in 3.x-dev? It greatly simplifies reasoning about this process if site's can't be concurrently processed.

Reply:
@diosmosis I just double checked and the same site might be processed concurrently in 3.x eg if you use `SharedSiteIds` or if you use `force-idsites=X`.

Reply:
Is it complicated to add @diosmosis ? It would certainly prevent possible random issues of not rearchived data having this there I would say

Reply:
Added it + a fix that the test picked up.

Reply:
@diosmosis did you push the fix? Only seeing 1 commit so far

Reply:
@tsteur pushed to the wrong branch, should be here now